python/ray/data/_internal/logical/optimizers.py

```diff
@@ -19,30 +30,211 @@ class PhysicalOptimizer(Optimizer):
+              the same class AND constructor args are the same for both.
+            * They have compatible remote arguments.
+        """
+        from ray.data._internal.execution.operators.map_operator import MapOperator
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        # We only support fusing MapOperators.
+        if not isinstance(down_op, MapOperator) or not isinstance(up_op, MapOperator):
+            return False
+
+        down_logical_op = self._op_map[down_op]
+        up_logical_op = self._op_map[up_op]
+
+        # Allow fusing tasks->actors if the resources are compatible (read->map), but
+        # not the other way around. The latter (downstream op) will be used as the
+        # compute if fused.
+        if (
+            is_task_compute(down_logical_op._compute)
+            and isinstance(up_logical_op, AbstractMap)
+            and get_compute(up_logical_op._compute)
+            != get_compute(down_logical_op._compute)
+        ):
+            return False
+
+        # Fusing callable classes is only supported if they are the same function AND
+        # their construction arguments are the same.
+        # TODO(Clark): Support multiple callable classes instantiating in the same actor
+        # worker.
+        if (
+            isinstance(down_logical_op._fn, CallableClass)
+            and isinstance(up_logical_op, AbstractMap)
+            and isinstance(up_logical_op._fn, CallableClass)
+            and (
+                up_logical_op._fn != down_logical_op._fn
+                or (
+                    up_logical_op._fn_constructor_args
+                    != down_logical_op._fn_constructor_args
+                    or up_logical_op._fn_constructor_kwargs
+                    != down_logical_op._fn_constructor_kwargs
+                )
+            )
+        ):
+            return False
+
+        # Only fuse if the ops' remote arguments are compatible.
+        if not _are_remote_args_compatible(
+            up_logical_op._ray_remote_args or {}, down_logical_op._ray_remote_args or {}
+        ):
+            return False
+
+        # Otherwise, ops are compatible for fusion.
+        return True
+
+    def _fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator):
+        """Fuse the downstream operator with its upstream operator."""
+        from ray.data._internal.execution.operators.map_operator import MapOperator
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        # Fuse operator names.
+        name = up_op.name + "->" + down_op.name
+
+        down_logical_op = self._op_map.pop(down_op)
+        up_logical_op = self._op_map.pop(up_op)
+
+        # Merge target block sizes.
+        down_target_block_size = down_logical_op._target_block_size
+        up_target_block_size = (
+            up_logical_op._target_block_size
+            if isinstance(up_logical_op, AbstractMap)
+            else None
```

Comment:
We should check `down_logical_op` is `AbstractMap` before accessing `down_logical_op._target_block_size`, right?

Reply:
Yep, see my above comment about allow-listing downstream and upstream logical ops!